src/pages/workspace/WorkspaceNewRoomPage.js

```diff
@@ -76,19 +76,85 @@ class WorkspaceNewRoomPage extends React.Component {
+            return;
+        }
+        Report.createPolicyRoom(
+            this.state.policyID,
+            this.state.roomName,
+            this.state.visibility,
+        );
+    }
+
     /**
-     * Called when a workspace is selected.
-     * @param {String} policyID
+     * @returns {Boolean}
      */
-    onWorkspaceSelect(policyID) {
-        this.setState({policyID});
+    validate() {
+        const errors = {};
+
+        const isExistingRoomName = _.some(
+            _.values(this.props.reports),
+            report => report && report.policyID === this.state.policyID
+            && report.reportName === this.state.roomName,
+        );
+
+        // We error if the user doesn't enter a room name or left blank
+        if (!this.state.roomName || this.state.roomName === CONST.POLICY.ROOM_PREFIX) {
+            errors.roomName = this.props.translate('newRoomPage.pleaseEnterRoomName');
+        }
+
+        // We error if the room name already exists. We don't care if it matches the original name provided in this
+        // component because then we are not changing the room's name.
+        if (isExistingRoomName) {
+            errors.roomName = this.props.translate('newRoomPage.roomAlreadyExistsError');
+        }
+
+        // Certain names are reserved for default rooms and should not be used for policy rooms.
+        if (_.contains(CONST.REPORT.RESERVED_ROOM_NAMES, this.state.roomName)) {
+            errors.roomName = this.props.translate('newRoomPage.roomNameReservedError');
+        }
+
+        if (!this.state.policyID) {
+            errors.policyID = this.props.translate('newRoomPage.pleaseSelectWorkspace');
+        }
+
+        this.setState({errors});
+        return _.isEmpty(errors);
     }
 
     /**
-     * Called when the "Create Room" button is pressed.
+     * @param {String} inputKey
+     * @param {String} value
      */
-    onSubmit() {
-        Report.createPolicyRoom(this.state.policyID, this.state.roomName, this.state.visibility);
+    clearErrorAndSetValue(inputKey, value) {
+        this.setState(prevState => ({
+            [inputKey]: value,
+            errors: {
+                ...prevState.errors,
+                [inputKey]: '',
+            },
+        }));
+    }
+
+    /**
+     * Modifies the room name to follow our conventions:
+     * - Max length 80 characters
+     * - Cannot not include space or special characters, and we automatically apply an underscore for spaces
+     * - Must be lowercase
+     * @param {String} roomName
+     *
+     * @returns {String}
+     */
+    modifyRoomName(roomName) {
+        const modifiedRoomNameWithoutHash = roomName.substr(1)
+            .replace(/ /g, '_')
+            .replace(/[^a-zA-Z\d_]/g, '')
+            .substr(0, CONST.REPORT.MAX_ROOM_NAME_LENGTH)
+            .toLowerCase();
+        const modifiedRoomName = `${CONST.POLICY.ROOM_PREFIX}${modifiedRoomNameWithoutHash}`;
+
+        return modifiedRoomName;
```

Comment:
The same thing exists on RoomNameInput. Could you please dry this up?

Reply:
Sure